### PR TITLE
use logic as SetProperties; handle settings specially; update & fix test

### LINKED ISSUE #4690: Microsoft.Test.SetProperties is not working correctly
## Version

Current main 4677d4ec6c5bd3659ec834d4c175f3036208e363

## Describe the bug

Microsoft.Test.SetProperties is not working correctly.
* assignment.Property.Value, assignment.Value.Value are null here https://github.com/microsoft/botbuilder-dotnet/blob/4677d4ec6c5bd3659ec834d4c175f3036208e363/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs#L61

* Also settings scope is read only.

## To Reproduce
1. Set a break point inside OnTurnError here https://github.com/microsoft/botbuilder-dotnet/blob/4677d4ec6c5bd3659ec834d4c175f3036208e363/libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestScript.cs#L133
2. Debug tests\Microsoft.Bot.Builder.Dialogs.Adaptive.Tests\Tests\TestScriptTests\TestScriptTests_PropertyMock.test.dialog
3. It sends out error message but not captured in test.

Related PR https://github.com/microsoft/botbuilder-dotnet/pull/4665
I think we should at least rollback mocking settings part

### PR DESCRIPTION
**Pull request by xieofxie**
Fixes #4690

## Description

Settings needs another turn to take effect.

## Specific Changes

  - use logic as SetProperties;
  - handle settings specially;
  - update & fix test

### CODE REVIEW

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs [hunk 1]
```diff
@@ -52,9 +55,27 @@ public async override Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler
             {
                 await inspector((dc) =>
                 {
+                    var mockSettingsMiddleware = dc.Context.TurnState.Get<MockSettingsMiddleware>();
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs:58]: >mockSettingsMiddleware [](start = 24, length = 22)

I think the original design where there is a mocksettings property on the test script is better.  I was confused by calling it mockProperties.  SetProperties is definitely useful for setting dialog/conversation, but the settings really needs to be setup before the dialog and it doesn't make sense to set those using SetProperties.

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs (hunk 1 above)
Comment (xieofxie) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs:58]: Sure. I will keep them both

**Comment by xieofxie** [comment]
@chrimc62 OK, I merged previous prs: https://github.com/microsoft/botbuilder-dotnet/pull/4201, https://github.com/microsoft/botbuilder-dotnet/pull/4226 for the SettingMocks (follow UserTokenMocks) property.
Also add a new test for this: TestScriptTests_SettingMock

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs [hunk 2]
```diff
@@ -54,7 +57,21 @@ public async override Task ExecuteAsync(TestAdapter adapter, BotCallbackHandler
...
+                        }
+
+                        if (val != null)
+                        {
+                            value = JToken.FromObject(val).DeepClone();
+                        }
+
+                        value = value?.ReplaceJTokenRecursively(dc.State);
+                        var property = assignment.Property.GetValue(dc.State);
+                        dc.State.SetValue(property, value);
```
Comment (chrimc62) [inline libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs:74]: What was wrong with just calling SetValue?

Comment (xieofxie) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs:74]: I use the logic from Microsoft.SetProperties. ReplaceJTokenRecursively will handle object binding correctly and I updated the test case for that:
```
                {
                    "property": "user.object",
                    "value": {
                        "value": "${user.object.value} again"
                    }
                }
```

**Comment by xieofxie** [comment]
@chrimc62 Will you review the updates? Thanks~

File: libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs (hunk 2 above)
Comment (chrimc62) [reply libraries/Microsoft.Bot.Builder.Dialogs.Adaptive.Testing/TestActions/SetProperties.cs:74]: I'm still missing why you can't just do dc.State.SetValue(property, value) why do you need to deep clone and replace jtoken?  We don't need to do that anywhere else to use SetValue.

---
In reply to: [505218331](https://github.com/microsoft/botbuilder-dotnet/pull/4720#discussion_r505218331) [](ancestors = 505218331)

**Review by chrimc62 — APPROVED** [review]
:shipit:

**Comment by xieofxie** [comment]
Hi @chrimc62 Because of this concurrent exception: https://github.com/microsoft/botbuilder-dotnet/pull/4720/checks?check_run_id=1324359368

I made an update on SourceMap: use ConcurrentDictionary instead of Dictionary.

Should we keep this change? Thanks!

**Comment by mrivera-ms** [comment]
@xieofxie please resolve the conflicts.

**Comment by mrivera-ms** [comment]
@chrimc62 could you please review the changes? Thank you.

**Comment by xieofxie** [comment]
@mrivera-ms Sure. I updated

**Comment by xieofxie** [comment]
Strange error: Insert nuget link into ApiCompat results file.
Blocked

**Comment by mrivera-ms** [comment]
> Strange failure in: Insert nuget link into ApiCompat results file.
> Blocked

@xieofxie this error was an external issue that has now been fixed. Please resolve the conflicts so we run checks again and merge. Thank you.

**Comment by xieofxie** [comment]
@mrivera-ms sure, updated

**Comment by chrimc62** [comment]
Do not check in the tests.schema/.uischema.  You need to update to the current bf and then rerun tests.  Other than that it looks fine.

**Review by chrimc62 — CHANGES_REQUESTED** [review]
:clock1:

**Comment by xieofxie** [comment]
@chrimc62 Sure, removed. But even after I updated cli, it still changes a lot in schema files.. That's weird..

**Comment by xieofxie** [comment]
OK.. If I don't add them, tests in mac won't pass..

**Comment by chrimc62** [comment]
Your files are not right.  Can you do bf --version and bf plugs:list and send it to me?  I suspect you have an older version of bf-dialog in your cli.

---
In reply to: [740306212](https://github.com/microsoft/botbuilder-dotnet/pull/4720#issuecomment-740306212) [](ancestors = 740306212)

**Comment by xieofxie** [comment]
One comment: bf-dialog is included in bf-cli now, so don't need to install separately. If bf-dialog is not updated, try ` bf plugins:uninstall @microsoft/bf-dialog`, ` bf plugins:install @microsoft/bf-dialog`

**Review by chrimc62 — APPROVED** [review]
:shipit:

**Pull request merged**